nifi-extension-bundles/nifi-github-bundle/nifi-github-extensions/src/main/java/org/apache/nifi/github/GitHubRepositoryClient.java

```diff
@@ -0,0 +1,404 @@
+                return ghContent;
+            } catch (final FileNotFoundException fnf) {
+                throwPathOrBranchNotFound(resolvedPath, branch);
+                return null;
+            }
+        });
+    }
+
+    private String getResolvedPath(final String path) {
+        return repoPath == null ? path : repoPath + "/" + path;
+    }
+
+    private void throwPathOrBranchNotFound(final String path, final String branch) throws FlowRegistryException {
+        throw new FlowRegistryException("Path [" + path + "] or Branch [" + branch + "] not found");
+    }
+
+    private <T> T execute(final GHRequest<T> action) throws FlowRegistryException, IOException {
+        try {
+            return action.execute();
+        } catch (final FlowRegistryException | IOException e) {
+            LOGGER.error(e.getMessage(), e);
+            throw e;
+        } catch (final Exception e) {
+            LOGGER.error(e.getMessage(), e);
+            throw new FlowRegistryException(e.getMessage(), e);
+        }
+    }
+
+    /**
+     * Functional interface for making a request to GitHub which may throw IOException.
+     *
+     * @param <T> the result of the request
+     */
+    private interface GHRequest<T> {
+
+        T execute() throws IOException, FlowRegistryException;
+
+    }
+
+    /**
+     * @return a new builder
+     */
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    /**
+     * Builder for the repository client.
+     */
+    public static class Builder {
+
+        private String apiUrl;
+        private GitHubAuthenticationType authenticationType;
+        private String personalAccessToken;
+        private String appInstallationToken;
+        private String repoOwner;
+        private String repoName;
+        private String repoPath;
+
+        public Builder apiUrl(final String apiUrl) {
+            this.apiUrl = apiUrl;
+            return this;
+        }
+
+        public Builder authenticationType(final GitHubAuthenticationType authenticationType) {
+            this.authenticationType = authenticationType;
+            return this;
+        }
+
+        public Builder personalAccessToken(final String personalAccessToken) {
+            this.personalAccessToken = personalAccessToken;
+            return this;
+        }
+
+        public Builder appInstallationToken(final String appInstallationToken) {
+            this.appInstallationToken = appInstallationToken;
+            return this;
+        }
+
+        public Builder repoOwner(final String repoOwner) {
+            this.repoOwner = repoOwner;
+            return this;
+        }
+
+        public Builder repoName(final String repoName) {
+            this.repoName = repoName;
+            return this;
+        }
+
+        public Builder repoPath(final String repoPath) {
+            this.repoPath = repoPath;
+            return this;
+        }
+
+        public GitHubRepositoryClient build() throws IOException {
```

Comment:
Minor: if we consider some attributes mandatory, this can be a good place to validate if they are set.

Reply:
I don't think there is anything to validate here, the overall `GitHubFlowRegistryClient` should not pass validation if the required properties are not set. This lower level client will only be created when all the required valid properties are set.